pkg/reconciler/ingress/lister.go

```diff
@@ -0,0 +1,196 @@
+			}
+			for _, host := range hosts.List() {
+				for _, target := range targets {
+					url := fmt.Sprintf(target.urlTmpl, host)
+
+					if targets, existed := results[ip][target.targetPort]; existed {
+						targets.Insert(url)
+					} else {
+						results[ip][target.targetPort] = sets.NewString(url)
+					}
+				}
+			}
+		}
+	}
+	return results, nil
+}
+
+func (l *gatewayPodTargetLister) getGateway(name string) (*v1alpha3.Gateway, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Gateway name %q: %w", name, err)
+	}
+	if namespace == "" {
+		return nil, fmt.Errorf("unexpected unqualified Gateway name %q", name)
+	}
+	return l.gatewayLister.Gateways(namespace).Get(name)
+}
+
+// listGatewayPodsURLs returns a map where the keys are the Gateway Pod IPs and the values are the corresponding
+// URL templates and the Gateway Pod Port to be probed.
+func (l *gatewayPodTargetLister) listGatewayTargetsPerPods(gateway *v1alpha3.Gateway) (map[string][]probeTarget, error) {
+	selector := labels.SelectorFromSet(gateway.Spec.Selector)
+
+	services, err := l.serviceLister.List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list Services: %w", err)
+	}
+	if len(services) == 0 {
+		l.logger.Infof("Skipping Gateway %s/%s because it has no corresponding Service", gateway.Namespace, gateway.Name)
+		return nil, nil
+	}
+	service := services[0]
+
+	endpoints, err := l.endpointsLister.Endpoints(service.Namespace).Get(service.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Endpoints: %w", err)
+	}
+
+	targetsPerPods := make(map[string][]probeTarget)
+	for _, server := range gateway.Spec.Servers {
+		var urlTmpl string
+		switch server.Port.Protocol {
+		case "HTTP", "HTTP2":
+			if server.Tls == nil || !server.Tls.HttpsRedirect {
+				urlTmpl = "http://%%s:%d/"
+			} else {
+				continue
+			}
+		case "HTTPS":
+			urlTmpl = "https://%%s:%d/"
+		default:
+			l.logger.Infof("Skipping Server %q because protocol %q is not supported", server.Port.Name, server.Port.Protocol)
+			continue
+		}
+
+		portName, err := findNameForPortNumber(service, int32(server.Port.Number))
+		if err != nil {
+			l.logger.Infof("Skipping Server %q because Service %s/%s doesn't contain a port %d", server.Port.Name, service.Namespace, service.Name, server.Port.Number)
+			continue
+		}
+		for _, sub := range endpoints.Subsets {
+			portNumber, err := findPortNumberForName(sub, portName)
+			if err != nil {
+				l.logger.Infof("Skipping Subset %v because it doesn't contain a port %q", sub.Addresses, portName)
+				continue
+			}
+
+			for _, addr := range sub.Addresses {
+				targetsPerPods[addr.IP] = append(targetsPerPods[addr.IP], probeTarget{urlTmpl: fmt.Sprintf(urlTmpl, server.Port.Number), targetPort: strconv.Itoa(int(portNumber))})
+			}
+		}
+	}
+	return targetsPerPods, nil
+}
+
+// findNameForPortNumber finds the name for a given port as defined by a Service.
```

Comment:
Can we move these into `pkg/network` so that they can be shared?

Reply:
Done